common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -534,35 +515,33 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) {
     }
     AtomicReference<Map<Integer, AppShufflePartitionInfo>> shuffleMergePartitionsRef =
       new AtomicReference<>(null);
-    // Metadata of the determinate stage shuffle can be safely removed as part of finalizing
-    // shuffle merge. Currently once the shuffle is finalized for a determinate stages, retry
-    // stages of the same shuffle will have shuffle push disabled.
-    if (msg.shuffleMergeId == DETERMINATE_SHUFFLE_MERGE_ID) {
-      AppShuffleMergePartitionsInfo appShuffleMergePartitionsInfo =
-        appShuffleInfo.shuffles.remove(msg.shuffleId);
-      if (appShuffleMergePartitionsInfo != null) {
-        shuffleMergePartitionsRef.set(appShuffleMergePartitionsInfo.shuffleMergePartitions);
+    appShuffleInfo.shuffles.compute(msg.shuffleId, (shuffleId, mergePartitionsInfo) -> {
+      if (null == mergePartitionsInfo) {
+        //If the mergePartitions was never created then it means that there weren't any push
+        //blocks that were ever received for this shuffle. This could be the case when the driver
+        //doesn't wait for enough time to start the stage which reads this shuffle data.
+        return new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);
+      } else if (msg.shuffleMergeId < mergePartitionsInfo.shuffleMergeId
+          || mergePartitionsInfo.isFinalized()) {
+        throw new RuntimeException(
+            String.format("Shuffle merge finalize request for shuffle %s with" +
+                " shuffleMergeId %s is %s", msg.shuffleId, msg.shuffleMergeId,
+                ErrorHandler.BlockPushErrorHandler.STALE_SHUFFLE_FINALIZE_SUFFIX));
+      } else if (msg.shuffleMergeId > mergePartitionsInfo.shuffleMergeId) {
+        // If no blocks pushed for the finalizeShuffleMerge shuffleMergeId then return
+        // empty MergeStatuses but cleanup the older shuffleMergeId files.
+        mergedShuffleCleaner.execute(() ->
+            closeAndDeletePartitionFiles(mergePartitionsInfo.shuffleMergePartitions));
+        return new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);
+      } else {
+        // This block covers:
+        //  1. finalization of determinate stage
+        //  2. finalization of indeterminate stage if the shuffleMergeId related to it is the one
+        //  for which the message is received.
+        shuffleMergePartitionsRef.set(mergePartitionsInfo.shuffleMergePartitions);
+        return new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);
```

Comment:
nit: All paths in this lambda, which are not throwing exceptions, are doing the same thing - `new AppShuffleMergePartitionsInfo(msg.shuffleMergeId, true);`
Make that common ?